libraries/AP_Compass/AP_Compass_BMM350.cpp

```diff
@@ -0,0 +1,557 @@
+            power = 0;
+            break;
+    }
+
+    if (ret >= power) {
+        ret -= (power * 2);
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Read out OTP(one-time programmable memory) data of sensor which is the compensation coefficients
+ * @see https://github.com/boschsensortec/BMM350-SensorAPI
+ */
+bool AP_Compass_BMM350::read_otp_data()
+{
+    uint8_t data = 0;
+    uint8_t msb = 0;
+    uint8_t lsb = 0;
+    uint16_t otp_data[BMM350_OTP_DATA_LENGTH];
+
+    for (uint8_t index = 0; index < BMM350_OTP_DATA_LENGTH; index++) {
+        // Set OTP address
+        data = (BMM350_OTP_CMD_DIR_READ | index);
+        if (!_dev->write_register(BMM350_REG_OTP_CMD, data)) {
+            return false;
+        }
+
+        // Wait OTP status be ready
+        do {
+            hal.scheduler->delay_microseconds(300);
+            // Read OTP status
+            if (!read_bytes(BMM350_REG_OTP_STATUS, &data, 1) || ((data & BMM350_OTP_STATUS_ERROR_MASK) != 0)) {
+                return false;
+            }
+        } while (!(data & BMM350_OTP_STATUS_CMD_DONE));
+
+        // Read OTP data
+        if (!read_bytes(BMM350_REG_OTP_DATA_MSB, &msb, 1) ||
+            !read_bytes(BMM350_REG_OTP_DATA_LSB, &lsb, 1)) {
+            return false;
+        }
+
+        otp_data[index] = ((uint16_t)(msb << 8) | lsb) & 0xFFFF;
+    }
+
+    {
+        // Update magnetometer offset and sensitivity data.
+        uint16_t off_x_lsb_msb, off_y_lsb_msb, off_z_lsb_msb;
+        int8_t t_off;
+        int8_t sens_x, sens_y, sens_z, sens_t;
+        int8_t tco_x, tco_y, tco_z;
+        int8_t tcs_x, tcs_y, tcs_z;
+        int8_t cross_x_y, cross_y_x, cross_z_x, cross_z_y;
+
+        off_x_lsb_msb = otp_data[BMM350_MAG_OFFSET_X] & 0x0FFF;
+        off_y_lsb_msb = ((otp_data[BMM350_MAG_OFFSET_X] & 0xF000) >> 4) + (otp_data[BMM350_MAG_OFFSET_Y] & 0x00FF);
+        off_z_lsb_msb = (otp_data[BMM350_MAG_OFFSET_Y] & 0x0F00) + (otp_data[BMM350_MAG_OFFSET_Z] & 0x00FF);
+        t_off = int8_t(otp_data[BMM350_TEMP_OFF_SENS] & 0x00FF);
+
+        // 12-bit unsigned integer to be left-aligned in a 16-bit integer
+        mag_comp.dut_offset_coef.x = fix_sign(off_x_lsb_msb, BMM350_SIGNED_12_BIT);
+        mag_comp.dut_offset_coef.y = fix_sign(off_y_lsb_msb, BMM350_SIGNED_12_BIT);
+        mag_comp.dut_offset_coef.z = fix_sign(off_z_lsb_msb, BMM350_SIGNED_12_BIT);
+        mag_comp.dut_offset_coef.temp = t_off / 5.0f;
+
+        sens_x = (otp_data[BMM350_MAG_SENS_X] & 0xFF00) >> 8;
+        sens_y = (otp_data[BMM350_MAG_SENS_Y] & 0x00FF);
+        sens_z = (otp_data[BMM350_MAG_SENS_Z] & 0xFF00) >> 8;
+        sens_t = (otp_data[BMM350_TEMP_OFF_SENS] & 0xFF00) >> 8;
+
+        mag_comp.dut_sensit_coef.x = sens_x / 256.0f;
+        mag_comp.dut_sensit_coef.y = (sens_y / 256.0f) + BMM350_SENS_CORR_Y;
+        mag_comp.dut_sensit_coef.z = sens_z / 256.0f;
+        mag_comp.dut_sensit_coef.temp = sens_t / 512.0f;
+
+        tco_x = (otp_data[BMM350_MAG_TCO_X] & 0x00FF);
```

Comment:
you could probably just cast otp_data as uint8_t* and avoid a lot of this masking.